src/lv2/SurgeLv2Ui.cpp

```diff
@@ -0,0 +1,158 @@
+#include "SurgeLv2Ui.h"
+#include "SurgeLv2Util.h"
+#include "SurgeLv2Vstgui.h"
+#include "SurgeLv2Wrapper.h"
+#include "SurgeGUIEditor.h"
+#include <cassert>
+
+SurgeLv2Ui::SurgeLv2Ui(SurgeLv2Wrapper* instance,
+                       void* parentWindow,
+                       const LV2_URID_Map* uridMapper,
+                       const LV2UI_Resize* uiResizer,
+                       LV2UI_Write_Function writeFn,
+                       LV2UI_Controller controller)
+    : _editor(new SurgeGUIEditor(instance, instance->synthesizer(), this)), _instance(instance),
+      _writeFn(writeFn), _controller(controller)
+{
+   instance->setEditor(this);
+
+   _editor->open(parentWindow);
+
+   if (uiResizer)
+   {
+      VSTGUI::ERect* rect = nullptr;
+      _editor->getRect(&rect);
+      uiResizer->ui_resize(uiResizer->handle, rect->right - rect->left, rect->bottom - rect->top);
+   }
+}
+
+SurgeLv2Ui::~SurgeLv2Ui()
+{
+   _instance->setEditor(nullptr);
+   _editor->close();
+}
+
+void SurgeLv2Ui::setParameterAutomated(int externalparam, float value)
+{
+   _writeFn(_controller, externalparam, sizeof(float), 0, &value);
```

Comment:
you need to offset the audio and event ports here, no?

Reply:
I have order the control ports first on purpose, in order to avoid dealing with any offsets.

Reply:
ah, smart. :)